File: tree/binary_tree/0101_symmetric_tree.js
/*
Given the root of a binary tree, check whether it is 
a mirror of itself (i.e., symmetric around its center).
*/

// https://leetcode.com/problems/symmetric-tree/description/

// recursive solution
var isSymmetric = function(root) {
    var isMirror = function(left, right) {
        if (!left && !right) return true;
        if (!left || !right) return false;
        return (left.val === right.val) && isMirror(left.left, right.right) && isMirror(left.right, right.left);
    }
    if (!root) return true;
    return isMirror(root.left, root.right);

};

File: tree/traversal/0101_symmetric_tree.py
# Given the root of a binary tree, 
# check whether it is a mirror of itself (i.e., symmetric around its center).

# https://leetcode.com/problems/symmetric-tree/description/

# recursive
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        return self.isMirror(root.left, root.right)

    def isMirror(self, left: Optional[TreeNode], right: Optional[TreeNode]) -> bool:
        if not left and not right:
            return True
        if not left or not right:
            return False
        return (left.val == right.val) and self.isMirror(left.left, right.right) and self.isMirror(left.right, right.left)
    
